File: 1871-jump-game-vii/solution.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        n = len(s)
        if s[-1] == '1':
            return False

        # dp[i] is True if index i is reachable
        dp = [False] * n
        dp[0] = True

        reachable_count = 0

        for i in range(1, n):
            # Slide the window: add the element entering the window from the right
            if i >= minJump:
                if dp[i - minJump]:
                    reachable_count += 1

            # Slide the window: remove the element leaving the window from the left
            if i > maxJump:
                if dp[i - maxJump - 1]:
                    reachable_count -= 1

            # An index is reachable if s[i] is '0' and there is at least one reachable point in the window
            if s[i] == '0' and reachable_count > 0:
                dp[i] = True

        return dp[-1]

File: 1871-jump-game-vii/solution.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = s.length();
        if (s[n - 1] == '1') {
            return false;
        }

        // dp[i] will be true if index i is reachable
        vector<bool> dp(n, false);
        dp[0] = true;

        int reachable_count = 0;

        for (int i = 1; i < n; ++i) {
            // Add the newly entering element into the window [i - maxJump, i - minJump]
            if (i >= minJump) {
                if (dp[i - minJump]) {
                    reachable_count++;
                }
            }

            // Remove the element that just left the window from the left side
            if (i > maxJump) {
                if (dp[i - maxJump - 1]) {
                    reachable_count--;
                }
            }

            // If there is at least one reachable index in the window and current char is '0'
            if (s[i] == '0' && reachable_count > 0) {
                dp[i] = true;
            }
        }

        return dp[n - 1];
    }
};